### ISSUE TITLE
Make Language Server Supervision in the Project Manager More Robust

### ISSUE THREAD

**Issue by radeusgd** [enso-org/enso#1315]
### Summary

There are some possible improvements to make the supervisor more robust.

1. When a graceful stop of the Language Server is requested, it can fail to stop in a finite amount of time. We may consider if we want to have a timeout and kill the process if it does not terminate on time. The situation happens both when user requests to close the project and in case of restarting an unresponsive LS.

2. ~removed~

3. Boot timeout. The LanguageServerProcess correctly handles the boot timeout - it only counts the time that is needed to actually boot the server, not counting waiting on any locks. This logic is correctly handled up to the LanguageServerRegistry. However, LanguageServerGateway uses the `?` operator to ask the registry to start the server and this operation requires a timeout. Currently we just set it to a a longer time, but maybe we could find a better solution. This is not affected by installation time as the LanguageServerGateway assumes that the engine is already pre-installed and that is handled by the ProjectService and does not account into the timeouts at all.

The ideal solution would be to still allow a timeout (although a slightly larger one, to allow the user to react if the application is locked by other program), but finally time out the operation - *but* instead of sending a 'timed out' error, detect that the timeout is caused by a lock and return a 'resource busy' error. Moreover it must ensure that if the top-level request times out, the actual process start is cancelled or the process is terminated as soon as possible, to ensure that no stray Language Servers are running.

### Value

- The Project Manager is more resilient to Language Server failures.

### Specification

- [ ] Add a stop timeout for graceful stop and restarting.
- [ ] After exceeding this timeout, the process should be killed forcibly.
      If the process does not terminate after another timeout, we report that the stop / restart has failed and release resources related to that process (the thread and actors). As we could not kill the process not much more can be done - we hope that it will just die at some point.
- [ ] Add a separate start timeout to configuration (3).
- [ ] Monitor if lock notifications were sent and in such case replace the timed-out error with 'resource busy'.
- [ ] Ensure that, if the `start` action fails, no Language Server is left running.
  - If the server does not start within the desired timeout, ensure that it does not launch or at least that it terminates as soon as possible (for example simply by sending a Stop message to its controller).

### Acceptance Criteria & Test Cases

- The above specification is completed.
- [ ] Test cases were added for the described scenarios, if possible.
  - Especially important to test that if the startup times out on a lock, the process is terminated.

**Comment by radeusgd** [comment]
After a discussion with @4e6 we decided that (2), that is:

> 2. Currently only one restart is attempted at a time - if the LS fails to restart on first try, it is considered dead. (This does not account situations where the LS restarts successfully and starts running again and is restarted again after some time, such restarts can happen without limit.)
We may want to retry restarting. Restarting may fail for two reasons: the old instance may fail to stop (as described above) or the new one may fail to boot. The bootloader needs to track this state - a new instance should not be started before the old one is stopped, but if the first restart failed to reboot, there is no 'old' instance to stop.
Another issue with restarting is that if some other process grabs the ports that were used, all restart attempts will fail. With the current implementation we cannot change the ports as the clients are already expecting the specific ports after the first boot, so we just need to signal that the server has died.

is not really that useful because if the server fails to restart for the first time, it is very likely that a second restart will fail too, so it's better for now to just allow a single restart.

(1) and (3) were also updated.

**Comment by 4e6** [comment]
An unhandled message always shows up in the log (the last line) during the project manager shutdown sequence.
```
[info] [2021-05-03T08:28:06.379Z] [org.enso.projectmanager.boot.ProjectManager$] Stopping server...
[info] [2021-05-03T08:28:06.4Z] [org.enso.projectmanager.infrastructure.languageserver.LanguageServerKiller] Killing all servers
[info] [2021-05-03T08:28:17.632Z] [org.enso.projectmanager.infrastructure.languageserver.LanguageServerController] Language server shut down successfully [Project(c24e74ba-22e4-42b3-b151-f2a5173a9e89,Unnamed,UserProject,2021-04-14T05:57:10.384630Z,0.0.0-SNAPSHOT,Some(2021-05-03T08:27:47.815574Z),Some(/home/dbushev/enso/projects/Unnamed),Some(2021-04-14T05:57:10.391375Z))].
[info] [2021-05-03T08:28:17.635Z] [org.enso.projectmanager.infrastructure.languageserver.LanguageServerKiller] All language servers have been killed
[info] [2021-05-03T08:28:17.638Z] [akka.actor.LocalActorRef] Message [org.enso.projectmanager.infrastructure.languageserver.package$GracefulStop$] from Actor[akka://project-manager/user/language-server-registry/language-server-controller-c24e74ba-22e4-42b3-b151-f2a5173a9e89#-2106918054] to Actor[akka://project-manager/user/language-server-registry/language-server-controller-c24e74ba-22e4-42b3-b151-f2a5173a9e89/bootloader-language-server-c24e74ba-22e4-42b3-b151-f2a5173a9e89#1483013262] was not delivered. [1] dead letters encountered. If this is not an expected behavior then Actor[akka://project-manager/user/language-server-registry/language-server-controller-c24e74ba-22e4-42b3-b151-f2a5173a9e89/bootloader-language-server-c24e74ba-22e4-42b3-b151-f2a5173a9e89#1483013262] may have terminated unexpectedly. This logging can be turned off or adjusted with configuration settings 'akka.log-dead-letters' and 'akka.log-dead-letters-during-shutdown'.
```
The unhandled message itself is not critical, but debugging it revealed a related issue.

It happens when [LanguageServerController](https://github.com/enso-org/enso/blob/6060d31c793c814cb2d2453df4a45193ee6c1b92/lib/scala/project-manager/src/main/scala/org/enso/projectmanager/infrastructure/languageserver/LanguageServerController.scala#L264) receives the `ServerTermianted` message from the [LanguageServerBootloader](https://github.com/enso-org/enso/blob/6060d31c793c814cb2d2453df4a45193ee6c1b92/lib/scala/project-manager/src/main/scala/org/enso/projectmanager/infrastructure/languageserver/LanguageServerBootLoader.scala#L205-L210) and runs the [`stop()`](https://github.com/enso-org/enso/blob/6060d31c793c814cb2d2453df4a45193ee6c1b92/lib/scala/project-manager/src/main/scala/org/enso/projectmanager/infrastructure/languageserver/LanguageServerController.scala#L293) function. The function sends `GracefulStop` to the children and waits for them to terminate. 

1. Unhandled `GracefulStop` message. It is printed because `LanguageServerController` sends it to the `LanguageServerBootloader` child which is already stopped itself after sending the [`LanugageServerTerminated` message](https://github.com/enso-org/enso/blob/6060d31c793c814cb2d2453df4a45193ee6c1b92/lib/scala/project-manager/src/main/scala/org/enso/projectmanager/infrastructure/languageserver/LanguageServerBootLoader.scala#L205-L210). An actor knows that its child is stopped after handling the `Terminated` message. But in case of the `LanguageServerController` the following happens:
  `LanguageServerBootloader` sends `ServerTerminated` message to `LanguageServerController`
  `LanguageServerBootloader` stops itself
  `Terminated(languageServerBootloaderRef)` message is sent to `LanguageServerController`
  `LanguageServerController` receives `ServerTerminated` message and runs `stop`. `Terminated` message is not handed and waiting in the inbox, so the `LanguageServerController` doesn't know yet that the `LanguageServerBootloader` is stopped.
  `LanguageServerController` sends `GracefulStop` message to `LanguageServerBootloader` child (which is already stopped).
  `LanguageServerController` switches to `waitingForChildren` and handles `Terminated` message (one that was waiting in the inbox)
2. The logic of the function [`stop()`](https://github.com/enso-org/enso/blob/6060d31c793c814cb2d2453df4a45193ee6c1b92/lib/scala/project-manager/src/main/scala/org/enso/projectmanager/infrastructure/languageserver/LanguageServerController.scala#L293-L303) that sends `GracefulStop` to children and waits for `Terminate`. In the case of `LanguageServerBootloader` it relies on the fact that the language server is running and will be terminated successfully. It will be safer to introduce a new stage with a timeout. It should also fix (1) Unhandled message because the `LanguageServerControler` would have an opportunity to handle `Terminated` message from the `LanguageServerBootloader`
